core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+
+public abstract class BaseSessionCatalog implements SessionCatalog {
+  private final Cache<String, Catalog> catalogs = Caffeine.newBuilder()
+      .expireAfterAccess(10, TimeUnit.MINUTES)
+      .removalListener((id, catalog, cause) -> {
+        if (catalog instanceof Closeable) {
+          try {
+            ((Closeable) catalog).close();
```

Comment:
Does this mean any session is closed after 10 minutes of idle time? or that the session may not be closed but the Catalog will need to be reinitialized again?

Reply:
This closes the catalog, but we don't control anything about the session. We're just caching a catalog instance to use with a session. If we have to create a new one because it timed out, it should be initialized with the session details so nothing would change in the catalog's behavior.

Reply:
Can we add a debug logging statement here? This seems like something that might wind up being hard to debug.

Reply:
I've moved this interface and base class into its own PR, #4773.

I also removed this `close` call because it doesn't need to close. The `AsCatalog` adapter doesn't have any resources of its own. It's just a light wrapper to use a session context with the `SessionCatalog`.